### ISSUE TITLE
Matrix::variance() incorrect results when all elements are the same

### ISSUE THREAD

**Issue by bushkov** [dimforge/nalgebra#1218]
If all elements in a matrix are the same I would expect that `Matrix::variance()` would give a result close to 0.0. And it works for f64 values that don't have a fractional part:
```rust
    let v = DVector::repeat(10_000, 100000000.0);
    assert_eq!(v.variance(), 0.0)
```
But it gives results which are far from 0.0, once a fractional part is introduced:
```rust
    let v = DVector::repeat(10_000, 100000000.815068511);
    assert_eq!(v.variance(), 0.0)
```
```shell
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `2258.0`,
 right: `0.0`', src/main.rs:5:5
```

Simpler variance implementation gives more precise results which are closer to the true variance value of 0.0:
```rust
    let v = DVector::repeat(10_000, 100000000.815068511);
    let m = v.mean();
    let variance = v.fold(0.0, |acc, e| acc + (e - m) * (e - m)) / 10_000.0;
    assert_eq!(variance, 0.0)
```
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `2.1201040922846914e-11`,
 right: `0.0`', src/main.rs:7:5
```

**Comment by sebcrozet** [comment]
That makes sense. The current implementation is being more efficient since we only iterate through the vector once.
Changing it to a more numerically stable version sounds good to me.

**Comment by jamesneb** [comment]
I would like to try implementing this. Can anyone point me to the relevant sections of the codebase?

**Comment by vasilNnikolov** [comment]
I started fixing the issue, hopefully will be done with a first draft in 24 hours.

**Comment by bushkov** [comment]
Fixed in v0.32.3.
